include/swift/Basic/ExperimentalDependencies.h

```diff
@@ -0,0 +1,968 @@
+        implementation->getKey().isImplementation() &&
+        "Interface must be interface, implementation must be implementation.");
+  }
+
+  NodeT *getInterface() const { return interface; }
+  NodeT *getImplementation() const { return implementation; }
+
+  /// When creating an arc to represent a link from def to use, the use end of
+  /// the arc depends on if the dependency is a cascading one. Centralize that
+  /// choice here.
+  /// ("use" in the name represents the noun, not the verb.)
+  NodeT *useDependingOnCascading(bool ifCascades) {
+    return ifCascades ? interface : implementation;
+  }
+};
+
+//==============================================================================
+// MARK: DependencyKey
+//==============================================================================
+
+/// The dependency system loses some precision by lumping entities together for
+/// the sake of simplicity. In the future, it might be finer-grained. The
+/// DependencyKey carries the information needed to find all uses from a def
+/// because the data structures in the graph map the key of an entity to all the
+/// nodes representing uses of that entity, even though the node may not really
+/// use the entity. For example, argument names of functions are ignored.
+class DependencyKey {
+  NodeKind kind;
+  DeclAspect aspect;
+  /// The mangled context type name of the holder for \ref potentialMember, \ref
+  /// member, and \nominal kinds. Otherwise unused.
+  std::string context;
+  /// The basic name of the entity. Unused for \ref potentialMember and \ref
+  /// nominal kinds.
+  std::string name;
+
+public:
+  /// See \ref FrontendNode::FrontendNode().
+  DependencyKey()
+      : kind(NodeKind::kindCount), aspect(DeclAspect::aspectCount), context(),
+        name() {}
+
+  /// For constructing a key in the frontend.
+  DependencyKey(NodeKind kind, DeclAspect aspect, std::string context,
+                std::string name)
+      : kind(kind), aspect(aspect), context(context), name(name) {
+    assert(verify());
+  }
+
+  NodeKind getKind() const { return kind; }
+  DeclAspect getAspect() const { return aspect; }
+  StringRef getContext() const { return context; }
+  StringRef getName() const { return name; }
+
+  StringRef getSwiftDepsFromSourceFileProvide() const {
+    assert(getKind() == NodeKind::sourceFileProvide &&
+           "Receiver must be sourceFileProvide.");
+    return getName();
+  }
+
+  bool operator==(const DependencyKey &rhs) const {
+    return getKind() == rhs.getKind() && getAspect() == rhs.getAspect() &&
+           getContext() == rhs.getContext() && getName() == rhs.getName();
+  }
+
+  bool operator!=(const DependencyKey &rhs) const { return !(*this == rhs); }
+
+  size_t hash() const {
+    return std::hash<size_t>()(size_t(kind)) ^
+           std::hash<size_t>()(size_t(aspect) * size_t(NodeKind::kindCount)) ^
+           std::hash<std::string>()(name) ^ std::hash<std::string>()(context);
```

Comment:
Nitpick: `^` is a bad hash mixer. LLVM has a better, variadic hash function as `llvm::hash_combine`.

Reply:
Great pointer, thanks!